Add details about navigation nodes and reading node parameters (#138)
kyma-project/luigi: docs/navigation-configuration.md

- Use the [Luigi Client API](luigi-client-api.md) to access the node parameter values from the micro front-end. Use the `LuigiClient.getPathParams()` function. 
For example, to get the value of the sample project parameter, use `LuigiClient.getPathParams().projectId`. 
- Add a parameter to the context part of your configuration:
  ```
  {
    pathSegment: ':projectId',
    label: 'Project Details',
    viewUrl: 'https://my.microfrontend.com/projects/details.html#id=:projectId;',
    context: {
    	project: ':projectId'
    }
  } 
  ```
In all cases the parameter is automatically replaced by the real value.

### Node parameters

Node parameters can be used to build the **viewUrl** and passed to the micro front-end specified in the navigation node selected in the navigation path. 
You can specify them in the main application URL, similarly to URL query parameters with a specific prefix. The prefix is `~` by default, but you can reconfigure it using the global **nodeParamPrefix** setting. 

All parameters without the prefix are not passed to the micro front-end and are consumed by the main application. 

> A sample **viewUrl** `https://luigiexample.com/home/projects/pr23?~sorting=asc&~page=2` supports sorting and paging (with the page number as a numerical value) by introducing the **sort** and **page** node parameters.

Author: This was reworked a bit

> A sample **viewUrl** `https://luigiexample.com/home/projects/pr23?~sorting=asc&~page=2` supports sorting and paging (with the page number as a numerical value) by introducing the **sort** and **page** node parameters.

Reviewer: ```suggestion
A sample **viewUrl** `https://luigiexample.com/home/projects/pr23?~sorting=asc&~page=2` supports sorting and paging, with the page number as a numerical value, by introducing the **sort** and **page** node parameters.
```

Reviewer: (with the page number as a numerical value) -> isn't it obvious? I'd consider removing this info.

> The navigation structure with the project list view using such sample node parameters parameters

Reviewer: ```suggestion
The navigation structure with the project list view using sample node parameters looks as follows:
```

````
{
  navigation: {
    nodes: [
      {
        pathSegment: 'home',
        label: 'Home',
        viewUrl: 'https://my.microfrontend.com/',
        children: [The
          {
            pathSegment: 'projects',
            label: 'Projects',
            viewUrl: 'https://my.microfrontend.com/projects/list.html#pagenr={nodeParams.page};sort={nodeParams.sorting}',
            children: [
              {
                pathSegment: ':projectId',
                label: 'Project Details',
                viewUrl: 'https://my.microfrontend.com/projects/details.html#id=:projectId;'
              }            
			]
          }
        ]
      }
    ]
  }
} 

````

>  Use the following options to work with path parameters:

Reviewer: ```suggestion
 Use the following options to work with node parameters:
```

Reviewer: ?

> - Build the **viewUrl** by placing them anywhere in the **viewUrl** value using the following syntax: `{nodeParams.<node param name>}`. For example, if the main application URL is `https://luigiexample.com/home/projects/?~sorting=asc&~page=2` then the **viewUrl** of a micro front-end would be `https://my.microfrontend.com/projects/list.html#pagenr=2;sort=asc`.

Reviewer: ```suggestion
- Build the **viewUrl** by placing them anywhere in the **viewUrl** value using the following syntax: `nodeParams.{node param name}`. For example, if the main application URL is `https://luigiexample.com/home/projects/?~sorting=asc&~page=2` then the **viewUrl** of a micro front-end is `https://my.microfrontend.com/projects/list.html#pagenr=2;sort=asc`.
```

- Use the [Luigi Client API](luigi-client-api.md) to access the node parameter values from within the micro front-end. Use the `LuigiClient.getNodeParams()` function. 
For example, to get the value of the sorting parameter, use `LuigiClient.getNodeParams().sorting`. 

> # Navigation configuration parameters reference 

Reviewer: ```suggestion
## Navigation configuration parameters reference 
```

> You can use the listed parameters and functions to configure your navigation structure. The examples show you how you can use selected options.

Reviewer: ```suggestion
You can use the listed parameters and functions to configure your navigation structure. The examples show how to use the selected options.
```

Reviewer: ?

> ## Routing

Reviewer: ```suggestion
### Routing
```

You can configure the way Luigi tackles routing in your application in the **Routing** section of the configuration file. In the example, you can choose the routing strategy to apply in your application as either hash or path location routing.

- **useHashRouting** defines either hash-based (`url.com/#/yourpath`) or path-based (`url.com/yourpath`) routing.

> - **nodeParamPrefix** sets the prefix character when using the `LuigiClient.linkManager().withParam()` function, which provides a way to simply attach query parameters to the view URL for activities such as sorting and filtering.  The URL contains the parameters to allow deep linking. If you want to use a different character prefix, define yours here. The default character is `~`.

Reviewer: ```suggestion
- **nodeParamPrefix** sets the prefix character when using the `LuigiClient.linkManager().withParam()` function, which provides a way to simply attach query parameters to the view URL for activities such as sorting and filtering. The URL contains the parameters to allow deep linking. If you want to use a different character prefix, define yours here. The default character is `~`.
```

> ## Node navigation parameters

Reviewer: ```suggestion
### Node navigation parameters
```

> 

Reviewer: Add an introductory sentence:
`The node navigation parameters are as follows:`

- **nodeAccessibilityResolver** allows you to define a permission checker function that gets executed on every node. If it returns `false`, Luigi removes the node and its children from the navigation structure.

> - **nodeAccessibilityResolver** receives all values defined in the node configuration. See [angular sampleconfig.js](../core/examples/luigi-sample-angular/src/assets/sampleconfig.js) for the **constraints** example.

Reviewer: Broken link - please fix.

Reviewer: ?

## Node parameters

> 

Reviewer: `The node parameters are as follows:`

Reviewer: ?

- **pathSegment** specifies the partial URL of the current segment. **pathSegment** must not contain slashes.

>   - A static settings example reflects `luigidomain.test/settings`, 

Reviewer: ```suggestion
  - A static settings example reflects `luigidomain.test/settings`.
```

  - A dynamic settings example, prefixed with a colon, loads on any other value. 
- **label** contains the display name of the navigation node.
- **hideFromNav** shows or hides a navigation node. You can still navigate to the node but it does not show up in the top or left pane.

> - **viewUrl** contains the URL or path to a view rendered when clicking the navigation node. Use either a full URL or a relative path. This value may consist of variables if you have specified a **navigationContext** with a dynamic **pathSegment**. If **viewUrl** is undefined, Luigi activates the child node specified in **defaultChildNode**. When both **viewUrl** and **defaultChildNode** are undefined, Luigi opens the first child of the current node.

Reviewer: ```suggestion
- **viewUrl** contains the URL or path to a view which renders when you click the navigation node. Use either a full URL or a relative path. This value may consist of variables if you specified a **navigationContext** with a dynamic **pathSegment**. If **viewUrl** is undefined, Luigi activates the child node specified in **defaultChildNode**. When both **viewUrl** and **defaultChildNode** are undefined, Luigi opens the first child of the current node.
```

- **navigationContext** contains a named node that is mainly for use in combination with a dynamic **pathSegment** to start navigation from a dynamic node using ` LuigiClient.navigationManager().fromContext('contextname')`.

> - **context** sends the specified object as context to the view. Use this parameter in combination with the dynamic **pathSegment** to receive the context through the context listeners of **Luigi client**. This is an alternative to using the dynamic value in the **viewUrl**.

Reviewer: ```suggestion
- **context** sends the specified object as context to the view. Use this parameter in combination with the dynamic **pathSegment** to receive the context through the context listeners of **Luigi Client**. This is an alternative to using the dynamic value in the **viewUrl**.
```

````

You can also navigate directly from any other node:

````
LuigiClient.linkManager().withParam({sort: 'asc'}).navigate('/something/sample_1/products');
````

### Application path

The main application path is built from **pathSegment** values in the navigation path, joined with the **/** character.
The micro front-end view URL is the value of the **viewUrl** property of the last node in the navigation path.

The following example shows the structure of different navigation paths. If the URL of the main application is `https://luigiexample.com`, then:

- `https://luigiexample.com/home/projects/pr1` reflects the `home/projects/pr1` navigation path. It is a valid navigation path as it exists in the defined navigation tree structure. The micro front-end view URL rendered in the content area is `https://my.microfrontend.com/projects/details.html#id=pr1`.

- `https://luigiexample.com/home/maskopatol` defines an invalid `home/maskopatol` navigation path.

- `https://luigiexample.com/projects/pr1` defines the `projects/pr1` navigation path. It is not a valid navigation path, since a valid navigation path always starts from the root.

### Path parameters

> Use the path parameter values to define the **pathSegment** in your configuration. You can either use a static value for your **pathSegment**, or add a colon to this value as in `:projectId`, to and make it act as a parameter. This tells Luigi to accept any value for this **pathSegment** of the main application URL. The value replaces the parameter when it is further processed by the application.

Reviewer: ```suggestion
Use the path parameter values to define the **pathSegment** in your configuration. You can either use a static value for your **pathSegment** or add a colon to this value, as in `:projectId`, to make it act as a parameter. This tells Luigi to accept any value for this **pathSegment** of the main application URL. The value replaces the parameter when it is further processed by the application.
```

- The **viewUrl** of a micro front-end rendered in the content area of the application.

A sample navigation structure looks as follows:

````
{
  navigation: {
    nodes: [
      {
        pathSegment: 'home',
        label: 'Home',
        viewUrl: 'https://my.microfrontend.com/',
        children: [
          {
            pathSegment: 'settings',
            label: 'Settings',
            viewUrl: 'https://my.microfrontend.com/general/settings.html'
          },
          {
            pathSegment: 'projects',
            label: 'Projects',
            viewUrl: 'https://my.microfrontend.com/projects/list.html',
            children: [
              {
                pathSegment: 'pr1',
                label: 'Project one',
                viewUrl: 'https://my.microfrontend.com/projects/details.html#id=pr1'
              },
              {
                pathSegment: 'pr2',
                label: 'Project two',
                viewUrl: 'https://my.microfrontend.com/projects/details.html#id=pr2'
              }
            ]
          }
        ]
      }
    ]
  }
}
````
## Node navigation

> When you navigate between Nodes that are located on the same domain, Luigi triggers a hash or path change. Then it sends an updated context in order not to fully reload the view for a single-page application based micro front-end. Navigation between domains triggers a full page load in order to comply with cross-domain security concepts.

Reviewer 2: Nodes should be lower case

Author: Done - all mentions of "node" are now lowercase.

- Add a parameter to the context part of your configuration:
  ```
  {
    pathSegment: ':projectId',
    label: 'Project Details',
    viewUrl: 'https://my.microfrontend.com/projects/details.html#id=:projectId;',
    context: {
    	project: ':projectId'
    }
  } 
  ```
In all cases the parameter is automatically replaced by the real value.

### Node parameters

Node parameters can be used to build the **viewUrl** and pass to the micro front-end specified in the navigation Node selected in the navigation path. 
You can specify them in the main application URL, similarly to URL query parameters with a specific prefix. The prefix is `~` by default, but you can reconfigure it using the global **nodeParamPrefix** setting. 

All parameters without the prefix are not passed to the micro front-end and are consumed by the main application. 

The project list view of in the example supports sorting and paging (with the page number as a numerical value). You can introduce the following Node parameters to represent these values:
- sorting
- page

The main application URL with the parameters is as follows: `https://luigiexample.com/home/projects/pr23?~sorting=asc&~page=2`

> The navigation structure with the project list view using these the parametrs is as follows:

Reviewer 2: there seems to be something wrong with that sentence

Author: I reworked the structure a bit to make it clearer
